File: day09/day09.go
package main

import (
	_ "embed"
	"fmt"
	"math"
	"regexp"
	"slices"

	"github.com/neil-vass/advent-of-code-2025/shared/input"
	"github.com/neil-vass/advent-of-code-2025/shared/set"
)

//go:embed input.txt
var puzzleData string

func main() {
	lines := input.SplitIntoLines(puzzleData)
	fmt.Printf("Part 1: %d\n", SolvePart1(lines))
	fmt.Printf("Part 2: %d\n", SolvePart2(lines))
}

type Pos struct{ X, Y int }
type Rect struct{ Min, Max Pos }
type RectArea struct {
	Bounds Rect
	Area   int
}

type Polygon struct {
	Vertices    []Pos
	BoundingBox Rect
	Candidates  []RectArea
	Outside     set.Set[Pos]
}

func (poly Polygon) LineStaysInsidePolygon(line LineSegment) bool {
	for x := min(line.Start.X, line.End.X); x <= max(line.Start.X, line.End.X); x++ {
		for y := min(line.Start.Y, line.End.Y); y <= max(line.Start.Y, line.End.Y); y++ {
			if poly.Outside.Has(Pos{x, y}) {
				return false
			}
		}
	}
	return true
}

func (poly Polygon) BestRectangle() RectArea {
	for _, c := range poly.Candidates {
		corners := []Pos{
			c.Bounds.Min,
			{c.Bounds.Min.X, c.Bounds.Max.Y},
			c.Bounds.Max,
			{c.Bounds.Max.X, c.Bounds.Min.Y},
		}

		// Do edges stay inside?
		if !(poly.LineStaysInsidePolygon(LineSegment{corners[0], corners[1]}) &&
			poly.LineStaysInsidePolygon(LineSegment{corners[1], corners[2]}) &&
			poly.LineStaysInsidePolygon(LineSegment{corners[2], corners[3]}) &&
			poly.LineStaysInsidePolygon(LineSegment{corners[3], corners[0]})) {
			continue
		}

		// Passed all the tests!
		return c
	}

	panic("No suitable rectangles at all")
}

func SolvePart1(lines []string) int {
	corners := ParseTiles(lines)
	maxArea := 0
	for i := 0; i < len(corners)-1; i++ {
		for j := i + 1; j < len(corners); j++ {
			area := Area(corners[i], corners[j])
			if area > maxArea {
				maxArea = area
			}
		}
	}
	return maxArea
}

func SolvePart2(lines []string) int {
	poly := ParsePolygon(lines)
	winner := poly.BestRectangle()
	return winner.Area
}

var tileRe = regexp.MustCompile(`^(\d+),(\d+)$`)

func ParseTiles(lines []string) []Pos {
	tiles := []Pos{}
	for _, ln := range lines {
		var p Pos
		err := input.Parse(tileRe, ln, &p.X, &p.Y)
		if err != nil {
			panic(err) // Don't panic!
		}
		tiles = append(tiles, p)
	}
	return tiles
}

func ParsePolygon(lines []string) Polygon {
	tiles := []Pos{}
	minBound := Pos{math.MaxInt, math.MaxInt}
	maxBound := Pos{math.MinInt, math.MinInt}
	rectAreas := []RectArea{}

	for _, ln := range lines {
		var tile Pos
		err := input.Parse(tileRe, ln, &tile.X, &tile.Y)
		if err != nil {
			panic(err) // Don't panic!
		}

		// Build bounding box of polygon
		minBound.X = min(minBound.X, tile.X)
		minBound.Y = min(minBound.Y, tile.Y)
		maxBound.X = max(maxBound.X, tile.X)
		maxBound.Y = max(maxBound.Y, tile.Y)

		// Build list of rectangles we can choose from
		for _, otherTile := range tiles {
			minPos := Pos{
				X: min(tile.X, otherTile.X),
				Y: min(tile.Y, otherTile.Y),
			}
			maxPos := Pos{
				X: max(tile.X, otherTile.X),
				Y: max(tile.Y, otherTile.Y),
			}
			ra := RectArea{
				Bounds: Rect{minPos, maxPos},
				Area:   Area(tile, otherTile),
			}
			rectAreas = append(rectAreas, ra)
		}
		tiles = append(tiles, tile)
	}

	// Sort by area, greatest first
	slices.SortFunc(rectAreas, func(a, b RectArea) int { return b.Area - a.Area })

	poly := Polygon{
		Vertices:    append(tiles, tiles[0]),
		BoundingBox: Rect{minBound, maxBound},
		Candidates:  rectAreas,
	}

	// Make clockwise.
	signedArea := 0
	for i := range len(poly.Vertices) - 1 {
		from, to := poly.Vertices[i], poly.Vertices[i+1]
		signedArea += (from.X*to.Y - to.X*from.Y)
	}
	if signedArea > 0 {
		slices.Reverse(poly.Vertices)
	}

	// Mark the "you're stepping outside" border.
	poly.Outside = set.Set[Pos]{}
	for i := range len(poly.Vertices) - 1 {
		edge := LineSegment{poly.Vertices[i], poly.Vertices[i+1]}
		if edge.IsHorizontal() {
			if edge.Start.Y < edge.End.Y {
				x := edge.Start.X - 1
				for y := edge.Start.Y; y <= edge.End.Y; y++ {
					poly.Outside.Add(Pos{x, y})
				}
			} else {
				x := edge.Start.X + 1
				for y := edge.End.Y; y <= edge.Start.Y; y++ {
					poly.Outside.Add(Pos{x, y})
				}
			}
		} else {
			if edge.Start.X < edge.End.X {
				y := edge.Start.Y + 1
				for x := edge.Start.X; x <= edge.End.X; x++ {
					poly.Outside.Add(Pos{x, y})
				}
			} else {
				y := edge.Start.Y - 1
				for x := edge.End.X; x <= edge.Start.X; x++ {
					poly.Outside.Add(Pos{x, y})
				}
			}
		}
	}

	// Remove edges that got mistakenly marked as "outside" above.
	for i := range len(poly.Vertices) - 1 {
		edge := LineSegment{poly.Vertices[i], poly.Vertices[i+1]}
		if edge.IsHorizontal() {
			x := edge.Start.X
			for y := min(edge.Start.Y, edge.End.Y); y <= max(edge.Start.Y, edge.End.Y); y++ {
				delete(poly.Outside, Pos{x, y})
			}
		} else {
			y := edge.Start.Y
			for x := min(edge.Start.X, edge.End.X); x <= max(edge.Start.X, edge.End.X); x++ {
				delete(poly.Outside, Pos{x, y})
			}
		}
	}

	return poly
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func Area(a, b Pos) int {
	length := a.X - b.X
	if length < 0 {
		length = -length
	}
	length += 1
	height := a.Y - b.Y
	if height < 0 {
		height = -height
	}
	height += 1
	return length * height
}

type LineSegment struct {
	Start, End Pos
}

func (ls LineSegment) IsVertical() bool {
	return ls.Start.Y == ls.End.Y
}

func (ls LineSegment) IsHorizontal() bool {
	return ls.Start.X == ls.End.X
}

File: day09_python/src/day09.py

import math

import numpy as np

class Rectangle:
    def __init__(self, min_pos, max_pos):
        self.min_pos = min_pos
        self.max_pos = max_pos

def area(rect):
    return ((rect.max_pos[0] - rect.min_pos[0] + 1) *
            (rect.max_pos[1] - rect.min_pos[1] + 1))

class Polygon:
    def  __init__(self, vertices, min_bound, max_bound, candidates):
        self.vertices = vertices
        self.min_bound = min_bound
        self.max_bound = max_bound
        self.candidates = candidates
        self.mark_interior()
    
    def mark_interior(self):
        self.interior = np.zeros((self.max_bound[0]+1, self.max_bound[1]+1), dtype=bool)

        # Make clockwise.
        signed_area = 0
        for i in range(len(self.vertices) - 1):
            start, end = self.vertices[i], self.vertices[i+1]
            signed_area += (start[0]*end[1] - end[0]*start[1])
	
        if signed_area > 0:
            self.vertices.reverse()

        # Add edges
        for i in range(len(self.vertices) - 1):
            edge = (self.vertices[i], self.vertices[i+1])
            if edge[0][0] == edge[1][0]: # is_horizontal
                x = edge[0][0]
                y_start = min(edge[0][1], edge[1][1])
                y_end = max(edge[0][1], edge[1][1]) +1
                self.interior[x, y_start:y_end] = 1
            else:
                y = edge[0][1]
                x_start = min(edge[0][0], edge[1][0])
                x_end = max(edge[0][0], edge[1][0]) +1
                self.interior[x_start:x_end, y] = True

        # Flood fill.
        for i in range(len(self.vertices) - 1): 
            edge = (self.vertices[i], self.vertices[i+1])
            if edge[0][0] == edge[1][0]:
                if edge[0][1] < edge[1][1]:
                    for y in range(edge[0][1], edge[1][1]+1):
                        x_start = edge[0][0] +1
                        x_end = x_start + self.interior[x_start:, y].nonzero()[0][0]
                        self.interior[x_start:x_end, y] = True
				


    def best_rectangle(self):
        for i, c in enumerate(self.candidates):
            if i % 100 == 0:
                print (f'{i} of {len(self.candidates)}')
            if self.interior[c.min_pos[0]:c.max_pos[0]+1, c.min_pos[1]:c.max_pos[1]+1].all():
                return c
        raise ValueError("No suitable rectangles at all")
        

def parse_polygon(lines):
    tiles = []
    rectanges = []
    min_bound = [math.inf, math.inf]
    max_bound = [-math.inf, -math.inf]

    for ln in lines:
        tile = [int(n) for n in ln.split(',')]
        min_bound = [min(min_bound[0], tile[0]), min(min_bound[1], tile[1])]
        max_bound = [max(max_bound[0], tile[0]), max(max_bound[1], tile[1])]

        for otherTile in tiles:
            min_pos = [min(tile[0], otherTile[0]), min(tile[1], otherTile[1])]
            max_pos = [max(tile[0], otherTile[0]), max(tile[1], otherTile[1])]
            rectanges.append(Rectangle(min_pos, max_pos))
		
        tiles.append(tile)

    # Sort by area, greatest first
    rectanges.sort(key=area, reverse=True)

    # One more tile on the end to close the loop
    tiles.append(tiles[0])
    
    poly = Polygon(tiles, min_bound, max_bound, rectanges)

    return poly
    

def solve_part_1(lines): 
    poly = parse_polygon(lines)
    winner = poly.candidates[0]
    return area(winner)


def solve_part_2(lines):
	poly = parse_polygon(lines)
	winner = poly.best_rectangle()
	return area(winner)


def fetch_data(path):
    with open(path, 'r') as f:
        for ln in f:
            yield ln 

if __name__ == "__main__":
    lines = fetch_data("./input.txt")
    print(solve_part_2(lines))
